transport/rif_comms/exceptions.py

```diff
@@ -0,0 +1,116 @@
+from abc import ABC, abstractmethod
+from typing import Union
+
+from grpc import RpcError, StatusCode
+
+
+class ClientException(Exception, ABC):
+
+    @property
+    @abstractmethod
+    def get_code(self):
+        """
+        return the code property
+        """
+
+    @property
+    @abstractmethod
+    def get_message(self):
+        """
+        return the code property
+        """
+
+    def __str__(self):
+        message = super(ClientException, self).__str__()
+        return f"Code: {self.get_code}, Message: {message}"
+
+
+class InvalidArgumentException(ClientException):
+    def __init__(self, code, message):
+        self.code = code
+        self.message = message
+        super(InvalidArgumentException, self).__init__(message)
+
+    def get_code(self):
+        return self.code
+
+    def get_message(self):
+        return self.message
+
+
+class InternalException(ClientException):
+    def __init__(self, code, message):
+        self.code = code
+        self.message = message
+        super(InternalException, self).__init__(message)
+
+    def get_code(self):
+        return self.code
+
+    def get_message(self):
+        return self.message
+
+
+class NotFoundException(ClientException):
+    def __init__(self, code, message):
+        self.code = code
+        self.message = message
+        super(NotFoundException, self).__init__(message)
+
+    def get_code(self):
+        return self.code
+
+    def get_message(self):
+        return self.message
+
+
+class FailedPreconditionException(ClientException):
+    def __init__(self, code, message):
+        self.code = code
+        self.message = message
+        super(FailedPreconditionException, self).__init__(message)
+
+    def get_code(self):
+        return self.code
+
+    def get_message(self):
+        return self.message
+
+
+class TimeoutException(ClientException):
+    def __init__(self, code, message):
+        self.code = code
+        self.message = message
+        super(TimeoutException, self).__init__(message)
+
+    def get_code(self):
+        return self.code
+
+    def get_message(self):
+        return self.message
+
+
+EXCEPTION_MAPPING = {
+    StatusCode.DEADLINE_EXCEEDED: TimeoutException,
+    StatusCode.INTERNAL: InternalException,
+    StatusCode.INVALID_ARGUMENT: InvalidArgumentException,
+    StatusCode.NOT_FOUND: NotFoundException,
+    StatusCode.FAILED_PRECONDITION: FailedPreconditionException
+}
+
+
+def client_handled_operation(func):
+    def inner(client: "Client", *args, **kwargs):
+        try:
+            return func(client, *args, **kwargs)
+        except RpcError as error:
+            raise get_exception(error)
+    return inner
+
+
+def get_exception(rpc_error: RpcError) -> Union[RpcError, ClientException]:
```

Comment:
if this module (exceptions.py) will 

- define exceptions
- define the mapping grpc status to exception
- define the way to convert them

i think it should be a class named ExceptionHandler or something

or.. the decorator, the exception mapping and mapping function can be moved back to the client

basically i think this file has enought responsability to be consider as a exception handler or converter class

Reply:
About moving back i don't know because all these are exception related, but maybe we can add a class called exception handler yeah or something like that

Reply:
> if this module (exceptions.py) will
> 
>     * define exceptions
> 
>     * define the mapping grpc status to exception
> 
>     * define the way to convert them
> 
> 
> i think it should be a class named ExceptionHandler or something
> 
> or.. the decorator, the exception mapping and mapping function can be moved back to the client
> 
> basically i think this file has enough responsability to be consider as a exception handler or converter class

i don't think this should be a stand-alone utils function. these are _client_ exceptions, they belong in the client class. just make this a static function and we should be ok.

same goes for the deco

Reply:
> About moving back i don't know because all these are exception related, but maybe we can add a class called exception handler yeah or something like that

i am also OK with an exception handler class. just have this in _a_ class